src/main/java/com/github/maxopoly/BottledExperience/PlayerListener.java

```diff
@@ -23,35 +27,77 @@ public PlayerListener(BottledExperience plugin) {
 	@EventHandler
 	public void playerInteract(PlayerInteractEvent e) {
 		Block b = e.getClickedBlock();
+		
 		if (e.getAction() == Action.LEFT_CLICK_BLOCK && b != null && b.getType() == Material.ENCHANTMENT_TABLE) {
 			Player p = e.getPlayer();
+			
 			int totalExperience = computeCurrentXP(p);
-			if (p.getItemInHand() != null && p.getItemInHand().getType() == Material.GLASS_BOTTLE &&
-					totalExperience >= plugin.getXpPerBottle() ) {
-				ItemMap playerInv = new ItemMap(p.getInventory());
-				int bottles = playerInv.getAmount(new ItemStack( Material.GLASS_BOTTLE));
-				int xpavailable = totalExperience / plugin.getXpPerBottle();
-				int remove = Math.min(bottles, xpavailable);
-				log(p.getName() + "interacted with enchanting table, inventory has " + bottles + " bottles, " 
-						+ totalExperience + " total experience, enough to fill " + remove + " bottles");
-				if (remove > 0) {
-					int endXP = totalExperience - (remove * plugin.getXpPerBottle());
-					p.setLevel(0);
-					p.setExp(0f);
-					log("Set xp for " + p.getName() + " to " + endXP);
-					p.giveExp(endXP);
-					ItemMap removeMap = new ItemMap();
-					removeMap.addItemAmount( new ItemStack(Material.GLASS_BOTTLE), remove);
-					for (ItemStack is : removeMap.getItemStackRepresentation()) {
-						p.getInventory().removeItem(is);
-						is.setType(Material.EXP_BOTTLE);
-						p.getInventory().addItem(is);
-						log("Turned " + is.getAmount() + " bottles into xp bottles for " + p.getName());
-					}
-				}
+			
+			if (p.getItemInHand() != null
+					&& p.getItemInHand().getType() == Material.GLASS_BOTTLE
+					&& totalExperience >= plugin.getXpPerBottle()
+					) {
+				createXPBottles(p, totalExperience);
+			}
+		}
+	}
+	
+	private void createXPBottles(Player p, int totalExperience) {
+		ItemMap playerInv = new ItemMap(p.getInventory());
+		int bottles = playerInv.getAmount(new ItemStack( Material.GLASS_BOTTLE));
+		int xpavailable = totalExperience / plugin.getXpPerBottle();
+		int remove = Math.min(bottles, xpavailable);
+		
+		log(p.getName() + " interacted with enchanting table, inventory has " + bottles + " bottles, " 
+				+ totalExperience + " total experience, enough to fill " + remove + " bottles");
+		
+		if (remove == 0) {
+			return;
+		}
+		
+		boolean noSpace = false;
+		int xpBottleCount = 0;
+		ItemMap removeMap = new ItemMap();
+		
+		removeMap.addItemAmount( new ItemStack(Material.GLASS_BOTTLE), remove);
+		
+		for (ItemStack is : removeMap.getItemStackRepresentation()) {
+			p.getInventory().removeItem(is);
+			is.setType(Material.EXP_BOTTLE);
+			
+			HashMap<Integer, ItemStack> result = p.getInventory().addItem(is);
+			
+			if(result != null && result.size() > 0) {
+				is.setType(Material.GLASS_BOTTLE);
+				p.getInventory().addItem(is);
```

Comment:
We'll have to have bonkill try to exploit this. Spamclicking the enchanting table with a full inventory might produce some weird effects.